fix(router): `RouterLinkActive` should run CD when setting `isActive`

Manduro opened the pull request:
regression caused by #15943
closes #19934

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] n/a Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #19934


## What is the new behavior?

Works with OnPush change detection 🎉 

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Toxicable commented on packages/router/src/directives/router_link_active.ts:
```diff
@@ -122,6 +122,7 @@ export class RouterLinkActive implements OnChanges,
       const hasActiveLinks = this.hasActiveLinks();
       if (this.isActive !== hasActiveLinks) {
         (this as any).isActive = hasActiveLinks;
+        this.cdr.detectChanges();
```
This should be `markForCheck` instead

Manduro replied:
I believe `detectChanges` is correct. Check the test and the issue on the regression. Neither of these would be fixed with `markForCheck`. We want to trigger change detection on the directive and it's children, no need to mark ancestors to be checked.

borowski-9ld replied:
Hi. We approached the problem with isActive + OnPush in our project and after adding `markForCheck` it started to work properly.

jasonaden commented:
@Manduro Can you please comment on the above and confirm whether `markForCheck` fixes this issue? Also, rebase and force push back to your branch, which will re-run CI so we can get this one green.

Manduro commented:
@jasonaden `markForCheck` does seem to work in my use case, but I can't get an OnPush test to pass with it.

borowski-9ld commented on packages/router/test/regression_integration.spec.ts:
```diff
@@ -107,6 +107,34 @@ describe('Integration', () => {
+         const router: Router = TestBed.get(Router);
+         const fixture = createRoot(router, OnPushComponent);
+         router.navigateByUrl('/simple');
+         tick();
```
@Manduro - This test fails with markForCheck because change detection is not triggered in tests automatically, Use `advance(fixture);` instead of `tick();`

Manduro replied:
Thanks, I don't know how I missed that!

Manduro committed:
fix(router): `RouterLinkActive` should run CD when setting `isActive`

When using the routerLinkActive directive inside a component that is using ChangeDetectionStrategy.OnPush and lazy loaded module routes the routerLinkActive directive does not update after clicking a link to a lazy loaded route that has not already been loaded.

Also the OnPush nav component does not set routerLinkActive correctly when the default route loads, the non-OnPush nav component works fine.

regression caused by #15943
closes #19934

Manduro commented:
@jasonaden Adjusted PR with `markForCheck`. The test included in the PR works as expected and fails before the code change, while it passes after. I also rebased on latest master.

CI partially fails, but it seems unrelated to this PR.

atscott reviewed (approved):
FYI, I rebased this PR and fixed the public api test failure. The change itself looks correct and I also verified the added test _does_ fail if I remove the `markForCheck` line. 

reviewed-for: fw-router

Will need approval from public API members

atscott commented:
[Google presubmit](https://test.corp.google.com/ui#id=OCL:316690515:BASE:316700308:1592326165518:55713018) also looks good. Only failures are unrelated/pre-existing.

Manduro commented:
Thanks for the effort 👍

jelbourn reviewed (approved):
LGTM

reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
reviewed-for: public-api

alxhub reviewed (approved):
reviewed-for: public-api
